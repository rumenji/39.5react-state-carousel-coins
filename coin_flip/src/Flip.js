import { useState } from "react";
import Coin from "./Coin";



 function Flip({ photos, title }) {
  
  const random = () => (Math.random()>=0.5)? 1 : 0;
  const [currPhotoIdx, setCurrPhotoIdx] = useState(random);

  let currPhoto = photos[currPhotoIdx];

  const[coin, setCurrCoin] = useState(null);
  const[headsCount, setHeadsCount] = useState(0);
  const[tailsCount, setTailsCount] = useState(0);
  const[flipsCount, setFlipsCount] = useState(0);

  function flipCoin(){
    setCurrPhotoIdx(random);
    setCurrCoin(<Coin caption={currPhoto.caption} src={currPhoto.src} />);
    setFlipsCount(flipsCount + 1);
    currPhoto.caption === "Heads" ? setHeadsCount(headsCount+1) : setTailsCount(tailsCount+1);
  }

  return (
    <div className="FlipCoin">
      <h1>{title}</h1>
      <div className="FlipCoin-main">
        {coin}
        <button className="FlipCoin-Button" onClick={flipCoin}>Flip Me!</button>
        <p>Out of {flipsCount}, there have been {headsCount} heads and {tailsCount} tails.</p>
      </div>
    </div>
  );
}

export default Flip;
