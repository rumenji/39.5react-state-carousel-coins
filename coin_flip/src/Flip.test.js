import { render, fireEvent } from "@testing-library/react";
import Flip from "./Flip.js";
import TEST_IMAGES from "./_testCommon.js";

beforeEach(function() {
  jest
    .spyOn(Math, "random")
    .mockReturnValueOnce(0.25)
    .mockReturnValueOnce(0.75);
});

afterEach(function() {
  Math.random.mockRestore();
});


it("renders", function(){
  render(<Flip photos={TEST_IMAGES} title="snapshot testing"/>)
})

it("it matches flip snapshot", function(){
  const {asFragment} = render(<Flip photos={TEST_IMAGES} title="snapshot testing"/>);
  expect(asFragment()).toMatchSnapshot();
})

it("works when you click the flip button", function() {
  const { container } = render(
    <Flip
      photos={TEST_IMAGES}
      title="images for testing"
    />
  );
  // expect no coin image to show
  expect(
    container.querySelector('.coin-image')
  ).not.toBeInTheDocument();

  // show coin
  const flipButton = container.querySelector(".FlipCoin-Button");
  fireEvent.click(flipButton);

  // expect a coin image to show
  expect(
    container.querySelector("img[class='Coin-image']")
  ).toBeInTheDocument();
});

it("correct image is displayed", function(){
  const { container } = render(<Flip photos={TEST_IMAGES} title="snapshot testing"/>);

  // expect no coin image to show
  expect(
    container.querySelector('.coin-image')
  ).not.toBeInTheDocument();

  // first flip
  const flipButton = container.querySelector(".FlipCoin-Button");
  fireEvent.click(flipButton);

  // expect a coin image to show
  expect(
    container.querySelector("img[alt='Heads']")
  ).toBeInTheDocument();
  expect(
    container.querySelector("img[alt='Tails']")
  ).not.toBeInTheDocument();

  // second flip
  fireEvent.click(flipButton);

  // expect a coin image to show
  expect(
    container.querySelector("img[alt='Tails']")
  ).toBeInTheDocument();
  expect(
    container.querySelector("img[alt='Heads']")
  ).not.toBeInTheDocument();


})



