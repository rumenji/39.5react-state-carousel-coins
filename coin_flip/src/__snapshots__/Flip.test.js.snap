// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches flip snapshot 1`] = `
<DocumentFragment>
  <div
    class="FlipCoin"
  >
    <h1>
      snapshot testing
    </h1>
    <div
      class="FlipCoin-main"
    >
      <button
        class="FlipCoin-Button"
      >
        Flip Me!
      </button>
      <p>
        Out of 0, there have been 0 heads and 0 tails.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
